### PR TITLE
Get default push remote from configuration

### PR DESCRIPTION
**Pull request by cruessler**
This PR is in response to #1093. It changes how GitUI gets the remote to push to. It does not change how it gets the remote to pull from, so it doesn’t directly address #1093, but it is a first step towards making the code paths for getting push and pull configuration independent.

It changes the following:

- It adds `get_default_remote_for_push_in_repo` which reads git configuration related to remote repositories. It first reads `branch.<name>.pushRemote`, then `remote.pushDefault`, then `branch.<name>.remote`. This is [according to the docs][git-config-remote-push-default].

I followed the checklist:

- [x] I added unittests
- [x] I ran `make check` without errors
- [x] I tested the overall application
- [x] I added an appropriate item to the changelog

[git-config-remote-push-default]: https://git-scm.com/docs/git-config#Documentation/git-config.txt-remotepushDefault

Test push behavior in status tab:

- [x] no upstream branch configured
- [x] upstream branch configured, `branch.<>.pushRemote` is set
- [x] upstream branch configured, `remote.pushDefault` is set
- [x] upstream branch configured, `branch.<>.remote` is set

### CODE REVIEW

File: asyncgit/src/sync/remotes/mod.rs [hunk 1]
```diff
@@ -51,6 +51,65 @@ pub fn get_default_remote(repo_path: &RepoPath) -> Result<String> {
...
+	repo: &Repository,
+) -> Result<Option<git2::Branch>> {
+	for b in repo.branches(None)? {
+		let branch = b?.0;
+		if branch.is_head() {
+			return Ok(Some(branch));
+		}
+	}
+	Ok(None)
+}
```
Comment (cruessler) [inline asyncgit/src/sync/remotes/mod.rs:67]: This function was copied verbatim from here: https://github.com/cruessler/gitui/blob/4ff0db9b80b816cd69a300877228dbab2c0986a3/asyncgit/src/sync/reword.rs#L37

**Comment by cruessler** [comment]
In addition to `branch.<name>.pushRemote`, we should also try to read `branch.<name>.remote`. I’ll update the PR soon.

https://git-scm.com/docs/git-config#Documentation/git-config.txt-branchltnamegtremote

**Comment by extrawurst** [comment]
looks good! yeah we should really try to get a unittest on this. i think local remotes to push to should work with libgit. so we can probably just simulate this that way.

**Comment by cruessler** [comment]
> In addition to `branch.<name>.pushRemote`, we should also try to read `branch.<name>.remote`. I’ll update the PR soon.
> 
> https://git-scm.com/docs/git-config#Documentation/git-config.txt-branchltnamegtremote

I’ve updated the PR.

**Comment by extrawurst** [comment]
@cruessler what's the status here? can this be merged? is it ready in your opinion? if not what TODOs do you still see?

**Comment by cruessler** [comment]
> @cruessler what's the status here? can this be merged? is it ready in your opinion? if not what TODOs do you still see?

I found a way to start testing this PR. I’ll add a testing-related TODO list to the initial post soon-ish. The function I added is pretty solid I think, what’s left to do is wire it up in the right places and make sure all combinations of config values work as expected.

I didn‘t leave a more high-level comment related to the overall progress, thinking that my last comment was sufficient. Now I see that was not the case. :-)

**Comment by cruessler** [comment]
@extrawurst I updated the PR. I think it’s now ready for testing! I tested different configurations for pushing and did not encounter any issues. I don’t have a good feeling for what I might have accidentally broken, so my testing was focused on pushing.

**Comment by cruessler** [comment]
Also, there is now a bit of duplication, but I thought we could address it in the future, once we now this is the way to go.

File: asyncgit/src/sync/remotes/mod.rs [hunk 2]
```diff
@@ -299,9 +374,44 @@ mod tests {
 			]
 		);
 
-		let res =
-			get_default_remote_in_repo(&repo(repo_path).unwrap());
-		assert_eq!(res.is_err(), true);
-		assert!(matches!(res, Err(Error::NoDefaultRemoteFound)));
+		let default_push_remote = get_default_remote_in_repo(&repo);
```
Comment (martihomssoler) [inline asyncgit/src/sync/remotes/mod.rs:377]: why is this not `get_default_remote_for_push_in_repo(&repo)`? The other asserts match against the default push repo, except this one.

**Review by martihomssoler — COMMENTED** [review]
I would add back the `test_default_remote_inconclusive()` test. Since the one that you added checks for push defaults and not for the inconclusive case for fetch repos. Apart from that, it look good. :+1:

File: asyncgit/src/sync/remotes/mod.rs (hunk 2 above)
Comment (cruessler) [reply asyncgit/src/sync/remotes/mod.rs]: Good catch! Splitting the test into two, restoring the previous `test_default_remote_inconclusive` also fixed this inconsistency.

**Comment by cruessler** [comment]
I updated this PR’s description to reflect the fact that this only tackles getting push configuration yet. Given what I’ve learned about the subject, adding something similar for pull configuration should be rather straight-forward. I think it makes sense to create a follow-up PR, but I’ll be equally happy to add more functionality in this PR.

**Comment by extrawurst** [comment]
@cruessler please add a changelog entry and then we are good to go

**Comment by cruessler** [comment]
> No let’s wrap this one up and make a follow up for pull. Can you create that @cruessler ?

Will do, I’m aiming for sometime this weekend.

**Comment by cruessler** [comment]
> @cruessler please add a changelog entry and then we are good to go

Done. (I didn’t find a specific issue this PR fixes, there’s one for pulling which I’ll link to in the follow-up PR.)

**Pull request merged**